### DISCUSSION TITLE
Best practices for querying metadata - with `context.instance` or GraphQL API?

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#18323] [Q&A] [answered]
Hi Dagster Community, is there a convention/preference of:
• Using GraphQL directly to query the metadata, OR
• Using Dagster-Instance (or OpExecutionContext.instance) to run Python commands similar to GraphQL query?
I believe that some of the GraphQL query has a Python implementation, but not all:
```
graphQL -> Python equivalent for (instance: DagsterInstance)

runsOrError -> instance.get_runs()
assetOrError -> instance.get_asset_records()
wipeAsset -> instance.wipe_assets()
```

Is there a documentation which GraphQL query/mutation correspond to which Python function?
Is there a convention we should stick to (e.g. Will the Python equivalent be well-supported? Or should we stick to directly querying GraphQL for stability purposes?)

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1701139495611379?thread_ts=1701139495.611379&cid=C01U954MEER).*

**Comment by smackesey** [comment, marked as answer]
In general you should use `context.instance` if you are inside an op/asset/schedule/sensor execution function. The GQL API is typically for use in non-Dagster environments.
